Localization extensibility docs (#11704)
aspnet/AspNetCore.Docs: aspnetcore/fundamentals/localization-extensibility.md

---
title: Localization Extensibility
author: hishamco
description: Learn how to extend the localization APIs in ASP.NET Core apps.
ms.author: riande
ms.date: 03/27/2019
uid: fundamentals/localization-extensibility
---
# Localization Extensibility

By [Hisham Bin Ateya](https://github.com/hishamco)

This article lists the extensibility points on the localization APIs and provides instructions on how to extend ASP.NET Core app localization.

## Extensible Points in Localization APIs

> ASP.NET Core localization APIs are build from ground up to be extensible, this will allow the developers to customize the localization according their needs. For instance [OrchardCore](https://github.com/orchardCMS/OrchardCore/) takes the initiative to build a `POStringLocalizer` which described in detail [Portable Object localization](fundamentals/portable-object-localization) to use `PO` files to store the localization resources.

Reviewer: ```suggestion
ASP.NET Core localization APIs are built from the ground up to be extensible, this allows developers to customize the localization according to their needs. For instance [OrchardCore](https://github.com/orchardCMS/OrchardCore/) has a `POStringLocalizer` which describes in detail using [Portable Object localization](fundamentals/portable-object-localization) to use `PO` files to store localization resources.
```

> This article lists two extensibility points that localization APIs have, including localization culture providers and localization resources baking store.

Reviewer: ```suggestion
This article lists two extensibility points that localization APIs have: the localization culture provider and localization resources baking store.
```

What do you mean by "localization resources baking store"?

Author: The underlying storage that contains the localization resources

## Localization Culture Providers

> ASP.NET Core localization APIs came up with four providers that can determine the current culture of the web application:

Reviewer: ```suggestion
ASP.NET Core localization has four providers by default. These determine the current culture of executing request:
```

1. `QueryStringRequestCultureProvider`
2. `CookieRequestCultureProvider`
3. `AcceptLanguageHeaderRequestCultureProvider`
4. `CustomRequestCultureProvider`

> All the above providers are described in detail [Localization middleware](fundamentals/localization), but in case if non of these providers don't meet your needs, you can build a custom provider using one of the following:

Reviewer: ```suggestion
All the above providers are described in detail in the [Localization middleware](fundamentals/localization) documentation, but if none of these providers meet your needs you can build a custom provider using one of the following:
```

### Using `CustomRequestCultureProvider`

> [CustomRequestCultureProvider](/dotnet/api/microsoft.aspnetcore.localization.customrequestcultureprovider?view=aspnetcore-2.1) provides a custom `RequestCultureProvider` that using a simple delegate allows you to determine the current localization culture.

Reviewer: ```suggestion
[CustomRequestCultureProvider](/dotnet/api/microsoft.aspnetcore.localization.customrequestcultureprovider?view=aspnetcore-2.1) provides a custom `RequestCultureProvider` that uses a simple delegate to determine the current localization culture.
```

::: moniker range=">= aspnetcore-2.2"

```csharp
options.AddInitialRequestCultureProvider(new CustomRequestCultureProvider(async context =>
{
    var currentCulture = "en";
    var segments = context.Request.Path.Value.Split(new char[] { '/' }, StringSplitOptions.RemoveEmptyEntries);
    if (segments.Length > 1 && segments[0].Length == 2)
    {
        currentCulture = segments[0];
    }

    var requestCulture = new ProviderCultureResult(culture);
    
    return Task.FromResult(requestCulture);
}));
```
::: moniker-end

::: moniker range="< aspnetcore-2.2"

```csharp
options.RequestCultureProviders.Insert(0, new CustomRequestCultureProvider(async context =>
{
    var currentCulture = "en";
    var segments = context.Request.Path.Value.Split(new char[] { '/' }, StringSplitOptions.RemoveEmptyEntries);
    if (segments.Length > 1 && segments[0].Length == 2)
    {
        currentCulture = segments[0];
    }

    var requestCulture = new ProviderCultureResult(culture);
    
    return Task.FromResult(requestCulture);
}));
```
::: moniker-end

### Using a new implemetation of `RequestCultureProvider`

> As mentioned ealier there are four implementation of `RequestCultureProvider` out of the box, which determines the request culture information from various sources.

Reviewer: Just delete this line, no need to repeat this info.

> The developer can easily use this as extensible point to create a new implementation of `RequestCultureProvider` that determines the request culture information from custom source such as configuration file, database ... etc.

Reviewer: ```suggestion
The developer can also create a new implementation of `RequestCultureProvider` that determines the request culture information from a custom source such as configuration file, database ... etc.
```

```csharp
public class AppSettingsRequestCultureProvider : RequestCultureProvider
{
    public string CultureKey { get; set; } = "culture";

    public string UICultureKey { get; set; } = "ui-culture";

    public override Task<ProviderCultureResult> DetermineProviderCultureResult(HttpContext httpContext)
    {
        if (httpContext == null)
        {
            throw new ArgumentNullException();
        }

        var configuration = httpContext.RequestServices.GetService<IConfigurationRoot>();
        var culture = configuration[CultureKey];
        var uiCulture = configuration[UICultureKey];

        if (culture == null && uiCulture == null)
        {
            return Task.FromResult((ProviderCultureResult)null);
        }

        if (culture != null && uiCulture == null)
        {
            uiCulture = culture;
        }

        if (culture == null && uiCulture != null)
        {
            culture = uiCulture;
        }
        
        var providerResultCulture = new ProviderCultureResult(culture, uiCulture);

        return Task.FromResult(providerResultCulture);
    }
}
```

## Localization Resources

> ASP.NET Core localization APIs came up with `ResourceManagerStringLocalizer` which an implementation of `IStringLocalizer` that is using `resx` files as underlying storage for the localization resources.

Reviewer: ```suggestion
ASP.NET Core Localization provides `ResourceManagerStringLocalizer` which is an implementation of `IStringLocalizer` that uses `resx` files to store localization resources. But you are not limited to using `resx` files. By implementing `IStringLocalized` you can use any data-source you want.
```

> But you are not limited to use the `resx` files if it isn't fit your need. We Already mentioned ealier that [OrchardCore](https://github.com/orchardCMS/OrchardCore/) introduced `POStringLocalizer` which using `PO` instead, to support pluralization.

Reviewer: You can just delete this line.

> We have seen many of the community projects that extends `IStringLocalizer` to create another localizers such as [EFStringLocalizer](https://github.com/aspnet/Entropy/tree/master/samples/Localization.EntityFramework), [JsonStringLocalizer](https://github.com/hishamco/My.Extensions.Localization.Json), [SqlLocalizer](https://github.com/damienbod/AspNetCoreLocalization) .. etc.

Reviewer: ```suggestion
Here are some examples of projects which implement `IStringLocalizer`: [EFStringLocalizer](https://github.com/aspnet/Entropy/tree/master/samples/Localization.EntityFramework), [JsonStringLocalizer](https://github.com/hishamco/My.Extensions.Localization.Json), [SqlLocalizer](https://github.com/damienbod/AspNetCoreLocalization) .. etc.
```

I feel like listing these is a bit fraught. The Entropy sample for example is not actively built anymore, and likely hasn't reacted the API changes we made to `IStringLocalizer` in 3.0. If we do keep that section then I think the suggestion above makes it clear that these are examples of how to implement IStringLocalizer, not just a list of community projects.

Author: The second two are the community projects

---
title: Localization Extensibility
author: hishamco
description: Learn how to extend the localization APIs in ASP.NET Core apps.
ms.author: riande

> ms.date: 8/27/2019

Author: Why it's 27 AUG, not 27 MAY?

Reviewer 2: Because it's August, Not May

Author: I see.. but it's 3 Aug not 27
